### PR TITLE
Gradient generation

### PR DESCRIPTION
**Pull request by Murkat-git**
Commit containing some utils and method to optimize antibody sequence with gradient

### CODE REVIEW

File: utils/general.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+project_path = os.path.dirname(os.path.realpath(antiberty.__file__))
+trained_models_dir = os.path.join(project_path, 'trained_models')
+ANTIBERTY_CHECKPOINT_PATH = os.path.join(trained_models_dir, 'AntiBERTy_md_smooth')
+ANTIBERTY_VOCAB_FILE = os.path.join(trained_models_dir, 'vocab.txt')
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+def exists(x):
+    return x is not None
```
Comment (Nik9kin) [inline utils/general.py:30]: Зачем и импортируешь exists, и сам пишешь?

File: utils/general.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+    @staticmethod
+    def get_mask_from_sequence(sequence: str) -> torch.Tensor:
+        mask = torch.zeros((len(sequence), 5))
+        a = Chain(sequence, scheme="chothia")
+        for reg, positions in a.regions.items():
+            if reg.startswith("FR"):
+                continue
+            for pos, amino in positions.items():
+                mask[pos.number] = torch.ones(5)
```
Comment (Nik9kin) [inline utils/general.py]: Работает некорректно. Обсудили почему: pos.number это **не** то же самое, что 0-based index

File: utils/general.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        return mask
+
+    def to_pdb(self, filename: str):
+        assert self.coords is not None
```
Comment (Nik9kin) [inline utils/general.py]: assert-ов лучше избегать в зрелом коде, потому что они не дают понимания произошедшей ошибки.
```suggestion
        if self.coords is None:
            raise ValueError('no coordinates data')
```

File: utils/scorer.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    n2 = normed_cross(n1, CA_C, eps=eps)
+    n3 = normed_cross(n1, n2, eps=eps)
+    rot = torch.stack([n1, n2, n3], -1)
+
+    return CA, rot
+
+
+from igfold.utils.constants import *
+def exists(x):
+    return x is not None
```
Comment (Nik9kin) [inline utils/scorer.py]: Опять?)

File: optimizers/gradient.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    aligned_target = do_kabsch(
+        mobile=target,
+        stationary=preds.detach(),
+        align_mask=None,
+    )
+    mse = F.mse_loss(
+        preds,
+        aligned_target,
+        reduction='none',
+    ).mean(-1)
```
Comment (Nik9kin) [inline optimizers/gradient.py]:
```suggestion
    ).sum(-1)
```

File: optimizers/gradient.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+    mse = torch.sum(
+            mse * mask,
+            dim=-1,
+        ).sum() / mask.mean(-1).sum()
```
Comment (Nik9kin) [inline optimizers/gradient.py]:
```suggestion
    rmsd_sq = torch.sum(mse * mask, dim=(-2, -1)) / mask.mean(-1).sum()
```

File: optimizers/gradient.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    
+    return mse
```
Comment (Nik9kin) [inline optimizers/gradient.py]:
```suggestion
    return rmsd_sq
```

**Review by Nik9kin — CHANGES_REQUESTED** [review]

File: optimizers/gradient.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+                return_tensors="pt",
+                padding=True,
+        ).to(device)
+        self.attention_mask = self.tokenizer_out["attention_mask"].to(device)
+
+        self.logits = F.one_hot(self.tokenizer_out.input_ids, num_classes=25).float() * sub_infinity
+
+        # Preprocess logits
+        vh_cdr_mask = F.pad(b.vh_cdr_mask.max(dim=1).values, pad=(1, self.logits.size(dim=1) - b.vh_cdr_mask.size(dim=0) - 1), mode="constant", value=0)
+        self.logits[0][vh_cdr_mask==1] = self.logits[0][vh_cdr_mask==1].where(self.logits[0][vh_cdr_mask==1]!=0, torch.tensor([-np.inf], device=device))
```
Comment (Nik9kin) [inline optimizers/gradient.py:77]: Что здесь происходит? Я не могу понять. Зачем это нужно?

Comment (Nik9kin) [inline optimizers/gradient.py]: Что здесь происходит? Я не могу понять. Зачем это нужно?

File: utils/general.py (hunk 1 above)
Comment (Murkat-git) [reply utils/general.py:30]: Дань уважения коллегам

**Review by Nik9kin — APPROVED** [review]
